### PR TITLE
feat: add reputation miner + decolonized BountyPayout contract + buidler/waffle/typescript infra

### LINKED ISSUE #2: Create a simple reputation miner
# Bounty

Given that it is costly to use Colony contracts nowadays (200+ gwei gas price, > $100 in tx fees) and we use only a tiny subset of the Colony (root domain reputation + miner), I suggest to stop calling Colony and mine reputation ourselves using Transfer events.

## Scope

create a script that will calculate current reputation in the organization based on payouts:
- use Transfer events initiated by PaymentSplitter and Payout events of BountyPayout contracts. We only want to index DAI transfers from our multisigs/safes
- implement decaying logic of Colony: reputation is reduced every hour so that it halves every 90 days
- iterate till the numbers produced are making sense (e.g. compare against Colony data or treasury spreadsheet)

## Deliverables

- CLI/webapp that shows reputation to date per account
- all necessary deployments
- usage instructions

## Funding Circle

Meta Crew via General Circle

## Bounty Owner/Gardener

@troggy as **Process Engineer**

## Gain for the Role

ability to use PaymentSplitter or decolonized BountyPayout contract for bounty payouts (10x gas savings vs colonized BountyPayout contract)

## Roles
bounty gardener: @troggy / 0 DAI
bounty worker: @troggy / 8x%
bounty reviewer: name / 1x%

## Gardener checklist

- [x] I have attached one of the `size-` labels to this bounty
- [x] I have submitted this bounty via http://bounty.leapdao.org/viewform
- [ ] I have added this bounty to bounty board here https://github.com/orgs/leapdao/projects/6

### PR DESCRIPTION
**Pull request by troggy**
What is here:
- [x] Contract ported from leap-contracts
- [x] using Solidity 0.6.8
- [x] using typescript for scripts and tests
- [x]  using ethers instead of web3 (lean!)
- [x]  using Buidler instead of Truffle (allows console.log in contracts and other perks)
- [x] using Waffle for testing (plays well with buidler)
- [x] using eth-gas-reporter to measure gas costs per function call (TODO: fix the reporting)
- [x] implement simple reputation miner in typescript
- [ ] ~generate typings for contracts~

Closes #2

### CODE REVIEW

**Comment by johba37** [comment]
cool, never seen this before: 

Screen Shot 2020-08-18 at 5 12 18 PM

**Comment by johba37** [comment]
in the debug stream of subgraph i can find this:
```
Error querying traces error = Err(Transport("Unexpected response status code: 429 Too Many Requests")) from = 10684767 to = 10684767, component: BlockStream
```

would i need to run a local instance to get around it? 🤔

**Comment by troggy** [comment]
> would i need to run a local instance to get around it?

that's a good catch. Probably not all the entities got caught thus wrong numbers. We may need to run our own node on AWS — it is quite slow to index (several hours), so not sure it is feasible for local run.

**Comment by troggy** [comment]
found and fixes three errors which were leading to wrong results:
1. non-unique payment id for payouts within a single tx ([fix](https://github.com/leapdao/bounty-system/pull/1/commits/64d34b72457d8fd0e5f6124e4181d62be74c4d1b#diff-4a7df0b7ce9a0a0e9606c9afb21c5774R17))
2. incorrect address comparison when checking for token is DAI. `==` is the proper comparison for AssemblyScript which is mapping.ts is written in. ([fix](https://github.com/leapdao/bounty-system/pull/1/commits/64d34b72457d8fd0e5f6124e4181d62be74c4d1b#diff-4a7df0b7ce9a0a0e9606c9afb21c5774R42))
3. incorrect Safes' address match because of case sensitive comparsion. Fixed by downcasing safes' addresses ([fix](https://github.com/leapdao/bounty-system/pull/1/commits/64d34b72457d8fd0e5f6124e4181d62be74c4d1b#diff-4a7df0b7ce9a0a0e9606c9afb21c5774L7))

Verified on a small index. Now reindexing the whole subgraph — will take a few hours for new data to appear.

File: subgraph/src/mapping.ts [hunk 1]
```diff
@@ -0,0 +1,56 @@
+import { Payout as PayoutEvent } from "../generated/BountyPayout/BountyPayout"
+import { Payout, Payee } from "../generated/schema"
+import { SplitERC20Call } from "../generated/PaymentSplitter/PaymentSplitter"
+import { BigInt, Bytes } from '@graphprotocol/graph-ts'
+
+let payers = new Array<string>()
+payers.push("0x27f083b6ce372a260f71305350f1e3e66efdca34")
```
Comment (Clearwood) [inline subgraph/src/mapping.ts]: could you comment a description of what accounts these are?

**Review by Clearwood — COMMENTED** [review]
This looks great :+1: I really enjoy the interface, I am going to try to test it a bit better over the coming day but haven't found any faults so far.

**Review by Clearwood — APPROVED** [review]

**Review by Clearwood — APPROVED** [review]

**Pull request merged**